doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+11. [log](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/log). This is a utility package for logging messages, for example: `logger.Info(something)`. It doesn't require python and is not *ML code*, so we **don't** have to migrate this package to python. 
+
+13. [sql/codegen](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/sql/codegen). This package implements the mentioned *text substitution* to generate *submitter programs*. This package is the core reason of **mirrored classes/functions**: Every function/package in `sqlflow_submitter` has a corresponding function/package in `sql/codegen`.
+
+    After the migration to python, we'll have a new python package that directly calls functions in `sqlflow_submitter`. Therefore, we don't need *text substitution* in the proposed architecture. We should **remove this package** after the migration.
+
+14. [sql/codegen/attribute](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/attribute) provides **limited semantic checking** for the python classes/functions. As discussed above, It should be migrated to python.
+
+16. [sql](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/sql) dispatches `ir`s of SQLFlow statements to appropriate platforms (PAI e.g.) and ML engines (XGBoost e.g.), and calls the corresponding `codegen` package (`codegen/xgboost` e.g.) to generate the *submitter program* that calls corresponding python functions, and spawns a process to execute the program. The main component of this package is a visitor hierarchy that implements a double dispatch mechanism. The hierachy is composed of `struct`s that implement the `ir.Executor` interface.
+    In the proposed architecture, this package still have to spawn a python process, but the double dispatch mechanism calls python *ML code* and should be reimplemented in python. After the migration, this package should be renamed to `executor`.
+
+16. [sqlfs](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/sqlfs) wraps a database table into a file system interface and is required by the `model` package. We would migrate this package to python together with `model`.
+
+17. [table_writer](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/table_writer) is a utility package for rendering database tables for a UI. We **don't** have to migrate this package to python. It's required by the `step` package and should be move to the `step` directory.
+
+## The Proposed Architecture
+
+### The Proposed Golang Package Structure
+
+Based on the foregoing analysis, the directory structure SQLFlow core code under `sqlflow/pkg` should be:
```

Comment:
It seems that we don't have the `code_gen` module in either go or python package.

Reply:
Yes, the proposed architecture doesn't have to generate a python script.

Reply:
In the MVP phase, golang will start a `step` binary, the `step` binary calls the python executable module `sqlflow.execute step` that calls the `python platform.train/predict/evaluate` functions.
In the second phase, a workflow step should start `sqlflow.execute step` directly.

See https://github.com/sql-machine-learning/sqlflow/pull/2335#discussion_r431035131